contd: improve retry logic when container fails to start

![2020-09-22_3494x695_scrot](https://user-images.githubusercontent.com/303164/93847610-af5bfa00-fca7-11ea-9217-127bbd64e36a.png)

So what seems to have happened is the the 0-fs process providing the files for the container crashed:
```
2020-09-07T13:28:59.591Z: main I > mount ready 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x725e62]

goroutine 650422 [running]:
github.com/threefoldtech/0-fs/meta.(*Dir).Name(0x0, 0xc0033f8980, 0x5)
	/gopath/src/github.com/threefoldtech/0-fs/meta/dir.go:31 +0x22
github.com/threefoldtech/0-fs/meta.(*sqlStore).get(0xc000176a80, 0xc001d26eb0, 0x42, 0x10, 0x964300, 0xc0010e5c01, 0xc0033d1340)
	/gopath/src/github.com/threefoldtech/0-fs/meta/store.go:253 +0x1dd
github.com/threefoldtech/0-fs/meta.(*sqlStore).Get(0xc000176a80, 0xc001d26eb0, 0x42, 0xc0033d1340, 0x1, 0x1)
	/gopath/src/github.com/threefoldtech/0-fs/meta/store.go:267 +0x3f
github.com/threefoldtech/0-fs/rofs.(*filesystem).GetAttr(0xc000168bc0, 0xc001d26eb0, 0x42, 0xc002d84858, 0x0, 0x0)
	/gopath/src/github.com/threefoldtech/0-fs/rofs/rofs.go:69 +0x159
github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/pathfs.(*readonlyFileSystem).GetAttr(0xc0001ba240, 0xc001d26eb0, 0x42, 0xc002d84858, 0xc0039ac000, 0xc0010e5e78)
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/pathfs/readonlyfs.go:26 +0x51
github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/pathfs.(*pathInode).GetAttr(0xc00070d2f0, 0xc002d847c0, 0x0, 0x0, 0xc002d84858, 0x481a5c)
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/pathfs/pathfs.go:603 +0x7c
github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/nodefs.(*rawBridge).GetAttr(0xc000176db0, 0xc002d84840, 0xc002d847b0, 0x958b20)
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/nodefs/fsops.go:139 +0x88
github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse.doGetAttr(0xc0001d2000, 0xc002d846c0)
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/opcode.go:246 +0x64
github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse.(*Server).handleRequest(0xc0001d2000, 0xc002d846c0, 0xc002d846c0)
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/server.go:404 +0x30c
github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse.(*Server).loop(0xc0001d2000, 0xc002c59c01)
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/server.go:376 +0x14d
created by github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse.(*Server).readRequest
	/gopath/src/github.com/threefoldtech/0-fs/vendor/github.com/hanwen/go-fuse/fuse/server.go:284 +0x332
```

And thus the container also crashed, and since the workloads was not asked to be deleted, containerd just keep trying to restart the container, but fail because of the 0-fs process that is dead

--- a/cmds/contd/main.go
+++ b/cmds/contd/main.go
@@ -76,6 +76,9 @@ func main() {
 		log.Info().Msg("shutting down")
 	})
 
+	// start watching for events
+	go containerd.Watch(ctx)
+
 	if err := server.Run(ctx); err != nil && err != context.Canceled {
 		log.Fatal().Err(err).Msg("unexpected error")
 	}

--- a/cmds/contd/main.go
+++ b/cmds/contd/main.go
@@ -62,7 +62,12 @@ func main() {
 		log.Fatal().Msgf("fail to connect to message broker server: %v", err)
 	}
 
-	containerd := container.New(moduleRoot, containerdCon)
+	client, err := zbus.NewRedisClient(msgBrokerCon)
+	if err != nil {
+		log.Fatal().Msgf("fail to connect to message broker server: %v", err)
+	}
+
+	containerd := container.New(client, moduleRoot, containerdCon)
 
 	server.Register(zbus.ObjectID{Name: module, Version: "0.0.1"}, containerd)
 

--- a/cmds/provisiond/main.go
+++ b/cmds/provisiond/main.go
@@ -139,7 +139,7 @@ func main() {
 		ZbusCl:         zbusCl,
 	})
 
-	server.Register(zbus.ObjectID{Name: module, Version: "0.0.1"}, pkg.ProvisionMonitor(engine))
+	server.Register(zbus.ObjectID{Name: module, Version: "0.0.1"}, pkg.Provision(engine))
 
 	log.Info().
 		Str("broker", msgBrokerCon).

--- a/cmds/zui/prov.go
+++ b/cmds/zui/prov.go
@@ -34,7 +34,7 @@ func provisionRender(client zbus.Client, grid *ui.Grid, render *Flag) error {
 
 	ctx := context.Background()
 
-	monitor := stubs.NewProvisionMonitorStub(client)
+	monitor := stubs.NewProvisionStub(client)
 	counters, err := monitor.Counters(ctx)
 	if err != nil {
 		return errors.Wrap(err, "failed to start net monitor stream")

--- a/go.mod
+++ b/go.mod
@@ -53,6 +53,7 @@ require (
 	github.com/opencontainers/runc v0.1.1 // indirect
 	github.com/opencontainers/runtime-spec v1.0.1
 	github.com/opencontainers/selinux v1.5.2 // indirect
+	github.com/patrickmn/go-cache v2.1.0+incompatible
 	github.com/pkg/errors v0.9.1
 	github.com/robfig/cron/v3 v3.0.0
 	github.com/rs/zerolog v1.19.0

--- a/go.mod
+++ b/go.mod
@@ -9,6 +9,7 @@ require (
 	github.com/alexflint/go-filemutex v0.0.0-20171028004239-d358565f3c3f
 	github.com/asaskevich/govalidator v0.0.0-20200108200545-475eaeb16496 // indirect
 	github.com/blang/semver v3.5.1+incompatible
+	github.com/cenkalti/backoff v2.2.1+incompatible
 	github.com/cenkalti/backoff/v3 v3.2.2
 	github.com/containerd/cgroups v0.0.0-20200327175542-b44481373989
 	github.com/containerd/containerd v1.4.0-beta.1.0.20200615192441-ae2f3fdfd1a4
@@ -62,7 +63,7 @@ require (
 	github.com/stretchr/testify v1.6.1
 	github.com/syndtr/gocapability v0.0.0-20200815063812-42c35b437635 // indirect
 	github.com/termie/go-shutil v0.0.0-20140729215957-bcacb06fecae
-	github.com/threefoldtech/tfexplorer v0.4.1-0.20201008114938-260c3694ef48
+	github.com/threefoldtech/tfexplorer v0.4.1-0.20201014110255-653c7314b944
 	github.com/threefoldtech/zbus v0.1.3
 	github.com/urfave/cli v1.22.4
 	github.com/vishvananda/netlink v1.1.0

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -20,6 +20,7 @@ import (
 	"github.com/opencontainers/runtime-spec/specs-go"
 
 	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/namespaces"
 	"github.com/containerd/containerd/oci"
 	"github.com/containerd/containerd/runtime/restart"
@@ -43,6 +44,7 @@ const (
 	defaultCPU    = 1
 
 	failuresBeforeDestroy = 4
+	restartDelay          = 2 * time.Second
 )
 
 var (
@@ -162,7 +164,6 @@ func (c *Module) Run(ns string, data pkg.Container) (id pkg.ContainerID, err err
 		containerd.WithNewSpec(opts...),
 		// this ensure that the container/task will be restarted automatically
 		// if it gets killed for whatever reason (mostly OOM killer)
-		restart.WithStatus(containerd.Running),
 		restart.WithBinaryLogURI(binaryLogsShim, nil),
 	)
 
@@ -207,21 +208,6 @@ func (c *Module) Run(ns string, data pkg.Container) (id pkg.ContainerID, err err
 		return id, err
 	}
 
-	// setting external logger process
-	uri, err := url.Parse("binary://" + binaryLogsShim)
-	if err != nil {
-		log.Error().Err(err).Msg("log uri")
-		return id, err
-	}
-
-	log.Info().Str("loguri", uri.String()).Msg("external logging process")
-
-	task, err := container.NewTask(ctx, cio.LogURI(uri))
-	if err != nil {
-		log.Error().Err(err).Msg("logger new task")
-		return id, err
-	}
-
 	// set user defined endpoint stats
 	for _, l := range data.Stats {
 		switch l.Type {
@@ -240,12 +226,56 @@ func (c *Module) Run(ns string, data pkg.Container) (id pkg.ContainerID, err err
 		}
 	}
 
-	// call start on the task to execute the redis server
-	if err := task.Start(ctx); err != nil {
+	if err := c.ensureTask(ctx, container); err != nil {
 		return id, err
 	}
 
 	return pkg.ContainerID(container.ID()), nil
+}
+
+func (c *Module) ensureTask(ctx context.Context, container containerd.Container) error {
+	uri, err := url.Parse("binary://" + binaryLogsShim)
+	if err != nil {
+		return err
+	}
+
+	log.Info().Str("loguri", uri.String()).Msg("external logging process")
+	task, err := container.Task(ctx, nil)
+
+	if err != nil && !errdefs.IsNotFound(err) {
+		return err
+	} else if err == nil {
+		//task found, we have to stop that first
+		_, err := task.Delete(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	//and finally create a new task
+	task, err = container.NewTask(ctx, cio.LogURI(uri))
+	if err != nil {
+		return err
+	}
+
+	return task.Start(ctx)
+}
+
+func (c *Module) start(ns, id string) error {
+	client, err := containerd.New(c.containerd)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	ctx := namespaces.WithNamespace(context.Background(), ns)
+
+	container, err := client.LoadContainer(ctx, string(id))
+	if err != nil {
+		return err
+	}
+
+	return c.ensureTask(ctx, container)
 }
 
 // Inspect returns the detail about a running container
@@ -335,7 +365,7 @@ func (c *Module) List(ns string) ([]pkg.ContainerID, error) {
 	return ids, nil
 }
 
-// Deletes stops and remove a container
+// Delete stops and remove a container
 func (c *Module) Delete(ns string, id pkg.ContainerID) error {
 	client, err := containerd.New(c.containerd)
 	if err != nil {

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -82,13 +82,55 @@ func New(client zbus.Client, root string, containerd string) *Module {
 		containerd = containerdSock
 	}
 
-	return &Module{
+	module := &Module{
 		containerd: containerd,
 		root:       root,
 		client:     client,
 		// values are cached only for 1 minute. purge cache every 20 second
 		failures: cache.New(time.Minute, 20*time.Second),
 	}
+
+	if err := module.upgrade(); err != nil {
+		log.Error().Err(err).Msg("failed to update containers configurations")
+	}
+
+	return module
+}
+
+// upgrade containers configurations. we make sure that any configuration changes apply
+// to the running containers..
+func (c *Module) upgrade() error {
+	// We make sure that ALL containers have auto-restart disabled since this is now
+	// managed completely by the watcher.
+	client, err := containerd.New(c.containerd)
+	if err != nil {
+		return err
+	}
+
+	defer client.Close()
+
+	ctx := context.Background()
+	nss, err := client.NamespaceService().List(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, ns := range nss {
+		ctx := namespaces.WithNamespace(context.Background(), ns)
+		containers, err := client.Containers(ctx)
+		if err != nil {
+			log.Error().Err(err).Str("namespace", ns).Msg("failed to list containers")
+			continue
+		}
+
+		for _, container := range containers {
+			if err := container.Update(ctx, restart.WithNoRestarts); err != nil { // mark this container as perminant down. so the watcher
+				log.Warn().Err(err).Msg("failed to clear up restart task status, continuing anyways") // does not try to restart it again
+			}
+		}
+	}
+
+	return nil
 }
 
 // Run creates and starts a container

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -55,6 +55,10 @@ var (
 		"mqueue": {},
 		"sysfs":  {},
 	}
+
+	// a marker value we use in failure cache to let the watcher
+	// no that we don't want to try restarting this container
+	permenant = struct{}{}
 )
 
 var (
@@ -379,9 +383,9 @@ func (c *Module) Delete(ns string, id pkg.ContainerID) error {
 		return err
 	}
 
-	if err := container.Update(ctx, restart.WithNoRestarts); err != nil {
-		log.Warn().Err(err).Msg("failed to clear up restart task status, continuing anyways")
-	}
+	// mark this container as perminant down. so the watcher
+	// does not try to restart it again
+	c.failures.Set(string(id), permenant, cache.DefaultExpiration)
 
 	task, err := container.Task(ctx, nil)
 	if err == nil {

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -302,7 +302,7 @@ func (c *Module) Inspect(ns string, id pkg.ContainerID) (result pkg.Container, e
 }
 
 // ListNS list the name of all the container namespaces
-func (c *containerModule) ListNS() ([]string, error) {
+func (c *Module) ListNS() ([]string, error) {
 	client, err := containerd.New(c.containerd)
 	if err != nil {
 		return nil, err
@@ -313,7 +313,7 @@ func (c *containerModule) ListNS() ([]string, error) {
 }
 
 // List all the existing container IDs from a certain namespace ns
-func (c *containerModule) List(ns string) ([]pkg.ContainerID, error) {
+func (c *Module) List(ns string) ([]pkg.ContainerID, error) {
 	client, err := containerd.New(c.containerd)
 	if err != nil {
 		return nil, err

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -58,7 +58,7 @@ var (
 
 	// a marker value we use in failure cache to let the watcher
 	// no that we don't want to try restarting this container
-	permenant = struct{}{}
+	permanent = struct{}{}
 )
 
 var (
@@ -385,7 +385,7 @@ func (c *Module) Delete(ns string, id pkg.ContainerID) error {
 
 	// mark this container as perminant down. so the watcher
 	// does not try to restart it again
-	c.failures.Set(string(id), permenant, cache.DefaultExpiration)
+	c.failures.Set(string(id), permanent, cache.DefaultExpiration)
 
 	task, err := container.Task(ctx, nil)
 	if err == nil {

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -25,6 +25,7 @@ import (
 	"github.com/containerd/containerd/runtime/restart"
 	"github.com/google/shlex"
 	"github.com/patrickmn/go-cache"
+	"github.com/threefoldtech/zbus"
 	"github.com/threefoldtech/zos/pkg"
 	"github.com/threefoldtech/zos/pkg/container/logger"
 	"github.com/threefoldtech/zos/pkg/container/stats"
@@ -65,12 +66,12 @@ var (
 type Module struct {
 	containerd string
 	root       string
-
-	failures *cache.Cache
+	client     zbus.Client
+	failures   *cache.Cache
 }
 
 // New return an new pkg.ContainerModule
-func New(root string, containerd string) *Module {
+func New(client zbus.Client, root string, containerd string) *Module {
 	if len(containerd) == 0 {
 		containerd = containerdSock
 	}
@@ -78,6 +79,7 @@ func New(root string, containerd string) *Module {
 	return &Module{
 		containerd: containerd,
 		root:       root,
+		client:     client,
 		// values are cached only for 1 minute. purge cache every 20 second
 		failures: cache.New(time.Minute, 20*time.Second),
 	}

--- a/pkg/container/container.go
+++ b/pkg/container/container.go
@@ -24,6 +24,7 @@ import (
 	"github.com/containerd/containerd/oci"
 	"github.com/containerd/containerd/runtime/restart"
 	"github.com/google/shlex"
+	"github.com/patrickmn/go-cache"
 	"github.com/threefoldtech/zos/pkg"
 	"github.com/threefoldtech/zos/pkg/container/logger"
 	"github.com/threefoldtech/zos/pkg/container/stats"
@@ -39,6 +40,8 @@ const (
 const (
 	defaultMemory = 256 * 1024 * 1204 // 256MiB
 	defaultCPU    = 1
+
+	failuresBeforeDestroy = 4
 )
 
 var (
@@ -54,27 +57,34 @@ var (
 var (
 	// ErrEmptyRootFS is returned when RootFS field is empty when trying to create a container
 	ErrEmptyRootFS = errors.New("RootFS of the container creation data cannot be empty")
+
+	_ pkg.ContainerModule = (*Module)(nil)
 )
 
-type containerModule struct {
+// Module implements pkg.Module interface
+type Module struct {
 	containerd string
 	root       string
+
+	failures *cache.Cache
 }
 
 // New return an new pkg.ContainerModule
-func New(root string, containerd string) pkg.ContainerModule {
+func New(root string, containerd string) *Module {
 	if len(containerd) == 0 {
 		containerd = containerdSock
 	}
 
-	return &containerModule{
+	return &Module{
 		containerd: containerd,
 		root:       root,
+		// values are cached only for 1 minute. purge cache every 20 second
+		failures: cache.New(time.Minute, 20*time.Second),
 	}
 }
 
 // Run creates and starts a container
-func (c *containerModule) Run(ns string, data pkg.Container) (id pkg.ContainerID, err error) {
+func (c *Module) Run(ns string, data pkg.Container) (id pkg.ContainerID, err error) {
 	// create a new client connected to the default socket path for containerd
 	client, err := containerd.New(c.containerd)
 	if err != nil {
@@ -153,6 +163,7 @@ func (c *containerModule) Run(ns string, data pkg.Container) (id pkg.ContainerID
 		restart.WithStatus(containerd.Running),
 		restart.WithBinaryLogURI(binaryLogsShim, nil),
 	)
+
 	if err != nil {
 		return id, err
 	}
@@ -236,7 +247,7 @@ func (c *containerModule) Run(ns string, data pkg.Container) (id pkg.ContainerID
 }
 
 // Inspect returns the detail about a running container
-func (c *containerModule) Inspect(ns string, id pkg.ContainerID) (result pkg.Container, err error) {
+func (c *Module) Inspect(ns string, id pkg.ContainerID) (result pkg.Container, err error) {
 	client, err := containerd.New(c.containerd)
 	if err != nil {
 		return result, err
@@ -323,13 +334,12 @@ func (c *containerModule) List(ns string) ([]pkg.ContainerID, error) {
 }
 
 // Deletes stops and remove a container
-func (c *containerModule) Delete(ns string, id pkg.ContainerID) error {
+func (c *Module) Delete(ns string, id pkg.ContainerID) error {
 	client, err := containerd.New(c.containerd)
 	if err != nil {
 		return err
 	}
 	defer client.Close()
-
 	ctx := namespaces.WithNamespace(context.Background(), ns)
 
 	container, err := client.LoadContainer(ctx, string(id))
@@ -372,7 +382,7 @@ func (c *containerModule) Delete(ns string, id pkg.ContainerID) error {
 	return container.Delete(ctx)
 }
 
-func (c *containerModule) ensureNamespace(ctx context.Context, client *containerd.Client, namespace string) error {
+func (c *Module) ensureNamespace(ctx context.Context, client *containerd.Client, namespace string) error {
 	service := client.NamespaceService()
 	namespaces, err := service.List(ctx)
 	if err != nil {

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -26,7 +26,7 @@ func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
 		c.failures.Set(event.ContainerID, int(0), cache.DefaultExpiration)
 	}
 
-	if marker == permenant {
+	if marker == permanent {
 		// if the marker is permenant. it means that this container
 		// is being deleted. we don't need to take any more action here
 		// (don't try to restart or delete)

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -8,7 +8,7 @@ import (
 	"github.com/containerd/typeurl"
 	"github.com/patrickmn/go-cache"
 	"github.com/rs/zerolog/log"
-	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/stubs"
 )
 
 func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
@@ -35,12 +35,13 @@ func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
 	if count < failuresBeforeDestroy {
 		return
 	}
-	log.Debug().Msg("deleting container due to so many crashes")
-	if err := c.Delete(ns, pkg.ContainerID(event.ContainerID)); err != nil {
-		log.Error().Err(err).Msg("failed to delete container")
-	}
 
-	//TODO: report to provisiond
+	log.Debug().Msg("deleting container due to so many crashes")
+
+	stub := stubs.NewProvisionStub(c.client)
+	if err := stub.DecommissionCached(event.ContainerID, "deleting container due to so many crashes"); err != nil {
+		log.Error().Err(err).Msg("failed to decommission reservation")
+	}
 }
 
 func (c *Module) handleEvent(ns string, event interface{}) {

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -1,0 +1,102 @@
+package container
+
+import (
+	"context"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/events"
+	"github.com/containerd/typeurl"
+	"github.com/patrickmn/go-cache"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg"
+)
+
+func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
+
+	log := log.With().
+		Str("namespace", ns).
+		Str("container", event.ContainerID).Logger()
+
+	log.Debug().Msg("task exited")
+
+	if _, ok := c.failures.Get(event.ContainerID); !ok {
+		c.failures.Set(event.ContainerID, int(0), cache.DefaultExpiration)
+	}
+
+	count, err := c.failures.IncrementInt(event.ContainerID, 1)
+	if err != nil {
+		// this should never happen because we make sure value
+		// is set
+		panic(err)
+	}
+
+	log.Debug().Int("count", count).Msg("recorded stops")
+
+	if count < failuresBeforeDestroy {
+		return
+	}
+	log.Debug().Msg("deleting container due to so many crashes")
+	if err := c.Delete(ns, pkg.ContainerID(event.ContainerID)); err != nil {
+		log.Error().Err(err).Msg("failed to delete container")
+	}
+
+	//TODO: report to provisiond
+}
+
+func (c *Module) handleEvent(ns string, event interface{}) {
+	switch event := event.(type) {
+	case *events.TaskExit:
+		c.handlerEventTaskExit(ns, event)
+	default:
+		log.Debug().Msgf("unhandled event: %+v", event)
+	}
+}
+
+// watch method will start a connection, and register for events
+// once an event is received, it will be handled. and exit on
+// first error or in case context was cancelled.
+// the caller must make sure this is called again in case of an
+// error
+func (c *Module) watch(ctx context.Context) error {
+	client, err := containerd.New(c.containerd)
+	if err != nil {
+		return err
+	}
+
+	defer client.Close()
+	log.Debug().Str("containerd", c.containerd).Msg("subscribe to events from containerd")
+
+	source, errors := client.Subscribe(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case envelope := <-source:
+			event, err := typeurl.UnmarshalAny(envelope.Event)
+			if err != nil {
+				log.Error().Err(err).Msg("failed to unmarshal event envelope")
+				continue
+			}
+
+			c.handleEvent(envelope.Namespace, event)
+		case err := <-errors:
+			return err
+		}
+	}
+}
+
+// Watch start watching for events coming from containerd.
+// Blocks forever. caller need to run this in a go routine
+//
+// different events types are handled differently. Now, only
+// TaskExit event is handled.
+func (c *Module) Watch(ctx context.Context) {
+	for {
+		err := c.watch(ctx)
+		if err == nil {
+			break // end of context
+		}
+
+		log.Err(err).Msg("error while watching events from containerd")
+	}
+}

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -43,7 +43,7 @@ func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
 	}
 
 	if reason != nil {
-		log.Debug().Msg("deleting container due to so many crashes")
+		log.Debug().Err(reason).Msg("deleting container due to restart error")
 
 		stub := stubs.NewProvisionStub(c.client)
 		if err := stub.DecommissionCached(event.ContainerID, reason.Error()); err != nil {

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -2,6 +2,8 @@ package container
 
 import (
 	"context"
+	"fmt"
+	"time"
 
 	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/api/events"
@@ -12,7 +14,6 @@ import (
 )
 
 func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
-
 	log := log.With().
 		Str("namespace", ns).
 		Str("container", event.ContainerID).Logger()
@@ -32,22 +33,33 @@ func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
 
 	log.Debug().Int("count", count).Msg("recorded stops")
 
+	var reason error
 	if count < failuresBeforeDestroy {
-		return
+		log.Debug().Msg("trying to restart the container")
+		<-time.After(restartDelay) // wait for 2 seconds
+		reason = c.start(ns, event.ContainerID)
+	} else {
+		reason = fmt.Errorf("deleting container due to so many crashes")
 	}
 
-	log.Debug().Msg("deleting container due to so many crashes")
+	if reason != nil {
+		log.Debug().Msg("deleting container due to so many crashes")
 
-	stub := stubs.NewProvisionStub(c.client)
-	if err := stub.DecommissionCached(event.ContainerID, "deleting container due to so many crashes"); err != nil {
-		log.Error().Err(err).Msg("failed to decommission reservation")
+		stub := stubs.NewProvisionStub(c.client)
+		if err := stub.DecommissionCached(event.ContainerID, reason.Error()); err != nil {
+			log.Error().Err(err).Msg("failed to decommission reservation")
+		}
 	}
 }
 
 func (c *Module) handleEvent(ns string, event interface{}) {
 	switch event := event.(type) {
 	case *events.TaskExit:
-		c.handlerEventTaskExit(ns, event)
+		// we run this handler in a go routine because
+		// - we don't want the restarts to slow down the event stream processing
+		// - this method does not return any useful value anyway, so safe to run
+		//   it in the background.
+		go c.handlerEventTaskExit(ns, event)
 	default:
 		log.Debug().Msgf("unhandled event: %+v", event)
 	}

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -20,15 +20,26 @@ func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
 
 	log.Debug().Msg("task exited")
 
-	if _, ok := c.failures.Get(event.ContainerID); !ok {
+	marker, ok := c.failures.Get(event.ContainerID)
+	if !ok {
+		// no previous value. so this is the first failure
 		c.failures.Set(event.ContainerID, int(0), cache.DefaultExpiration)
+	}
+
+	if marker == permenant {
+		// if the marker is permenant. it means that this container
+		// is being deleted. we don't need to take any more action here
+		// (don't try to restart or delete)
+		log.Debug().Msg("permenant delete marker is set")
+		return
 	}
 
 	count, err := c.failures.IncrementInt(event.ContainerID, 1)
 	if err != nil {
 		// this should never happen because we make sure value
 		// is set
-		panic(err)
+		log.Error().Err(err).Msg("error while checking number of failures")
+		return
 	}
 
 	log.Debug().Int("count", count).Msg("recorded stops")

--- a/pkg/container/watch.go
+++ b/pkg/container/watch.go
@@ -27,10 +27,10 @@ func (c *Module) handlerEventTaskExit(ns string, event *events.TaskExit) {
 	}
 
 	if marker == permanent {
-		// if the marker is permenant. it means that this container
+		// if the marker is permanent. it means that this container
 		// is being deleted. we don't need to take any more action here
 		// (don't try to restart or delete)
-		log.Debug().Msg("permenant delete marker is set")
+		log.Debug().Msg("permanent delete marker is set")
 		return
 	}
 

--- a/pkg/monitor.go
+++ b/pkg/monitor.go
@@ -4,7 +4,6 @@ package pkg
 //go:generate zbusc -module monitor -version 0.0.1 -name system -package stubs github.com/threefoldtech/zos/pkg+SystemMonitor stubs/system_monitor_stub.go
 //go:generate zbusc -module monitor -version 0.0.1 -name host -package stubs github.com/threefoldtech/zos/pkg+HostMonitor stubs/host_monitor_stub.go
 //go:generate zbusc -module identityd -version 0.0.1 -name monitor -package stubs github.com/threefoldtech/zos/pkg+VersionMonitor stubs/version_monitor_stub.go
-//go:generate zbusc -module provision -version 0.0.1 -name provision -package stubs github.com/threefoldtech/zos/pkg+ProvisionMonitor stubs/provision_monitor_stub.go
 
 import (
 	"context"
@@ -96,19 +95,4 @@ type HostMonitor interface {
 // VersionMonitor interface (provided by identityd)
 type VersionMonitor interface {
 	Version(ctx context.Context) <-chan semver.Version
-}
-
-// ProvisionCounters struct
-type ProvisionCounters struct {
-	Container int64 `json:"container"`
-	Volume    int64 `jons:"volume"`
-	Network   int64 `json:"network"`
-	ZDB       int64 `json:"zdb"`
-	VM        int64 `json:"vm"`
-	Debug     int64 `json:"debug"`
-}
-
-// ProvisionMonitor interface
-type ProvisionMonitor interface {
-	Counters(ctx context.Context) <-chan ProvisionCounters
 }

--- a/pkg/provision.go
+++ b/pkg/provision.go
@@ -1,0 +1,21 @@
+package pkg
+
+//go:generate zbusc -module provision -version 0.0.1 -name provision -package stubs github.com/threefoldtech/zos/pkg+Provision stubs/provision_stub.go
+
+import "context"
+
+// ProvisionCounters struct
+type ProvisionCounters struct {
+	Container int64 `json:"container"`
+	Volume    int64 `jons:"volume"`
+	Network   int64 `json:"network"`
+	ZDB       int64 `json:"zdb"`
+	VM        int64 `json:"vm"`
+	Debug     int64 `json:"debug"`
+}
+
+// Provision interface
+type Provision interface {
+	Counters(ctx context.Context) <-chan ProvisionCounters
+	DecommissionCached(id string, reason string) error
+}

--- a/pkg/provision/engine.go
+++ b/pkg/provision/engine.go
@@ -9,6 +9,7 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/cenkalti/backoff/v3"
 	"github.com/jbenet/go-base58"
 	"github.com/threefoldtech/zbus"
 	"github.com/threefoldtech/zos/pkg"
@@ -336,11 +337,21 @@ func (e *Engine) DecommissionCached(id string, reason string) error {
 		return errors.Wrapf(err, "failed to build result object for reservation: %s", id)
 	}
 
-	if err := e.reply(ctx, result); err != nil {
+	if err := e.decommission(ctx, r); err != nil {
 		log.Error().Err(err).Msgf("failed to update reservation result with failure: %s", id)
 	}
 
-	return e.decommission(ctx, r)
+	bf := backoff.NewExponentialBackOff()
+	bf.MaxInterval = 10 * time.Second
+	bf.MaxElapsedTime = 1 * time.Minute
+
+	return backoff.Retry(func() error {
+		err := e.reply(ctx, result)
+		if err != nil {
+			log.Error().Err(err).Msgf("failed to update reservation result with failure: %s", id)
+		}
+		return err
+	}, bf)
 }
 
 func (e *Engine) buildResult(id string, typ ReservationType, err error, info interface{}) (*Result, error) {

--- a/pkg/provision/engine.go
+++ b/pkg/provision/engine.go
@@ -320,6 +320,29 @@ func (e *Engine) reply(ctx context.Context, result *Result) error {
 	return e.feedback.Feedback(e.nodeID, result)
 }
 
+// DecommissionCached is used by other module to ask provisiond that
+// a certain reservation is dead beyond repair and owner must be informed
+// the decommission method will take care to update the reservation instance
+// and also decommission the reservation normally
+func (e *Engine) DecommissionCached(id string, reason string) error {
+	r, err := e.cache.Get(id)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	result, err := e.buildResult(id, r.Type, fmt.Errorf(reason), nil)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build result object for reservation: %s", id)
+	}
+
+	if err := e.reply(ctx, result); err != nil {
+		log.Error().Err(err).Msgf("failed to update reservation result with failure: %s", id)
+	}
+
+	return e.decommission(ctx, r)
+}
+
 func (e *Engine) buildResult(id string, typ ReservationType, err error, info interface{}) (*Result, error) {
 	result := &Result{
 		Type:    typ,

--- a/pkg/provision/explorer/feedback.go
+++ b/pkg/provision/explorer/feedback.go
@@ -1,8 +1,11 @@
 package explorer
 
 import (
+	"errors"
 	"fmt"
+	"time"
 
+	"github.com/cenkalti/backoff"
 	"github.com/threefoldtech/tfexplorer/client"
 	"github.com/threefoldtech/tfexplorer/models/generated/directory"
 	"github.com/threefoldtech/zos/pkg/provision"
@@ -13,13 +16,20 @@ import (
 type Feedback struct {
 	client    *client.Client
 	converter provision.ResultConverterFunc
+
+	strategy backoff.BackOff
 }
 
 // NewFeedback creates an ExplorerFeedback
 func NewFeedback(client *client.Client, converter provision.ResultConverterFunc) *Feedback {
+	strategy := backoff.NewExponentialBackOff()
+	strategy.MaxInterval = 15 * time.Second
+	strategy.MaxElapsedTime = 3 * time.Minute
+
 	return &Feedback{
 		client:    client,
 		converter: converter,
+		strategy:  strategy,
 	}
 }
 
@@ -30,15 +40,42 @@ func (e *Feedback) Feedback(nodeID string, r *provision.Result) error {
 		return fmt.Errorf("failed to convert result into schema type: %w", err)
 	}
 
-	return e.client.Workloads.NodeWorkloadPutResult(nodeID, r.ID, *wr)
+	return backoff.Retry(func() error {
+		err := e.client.Workloads.NodeWorkloadPutResult(nodeID, r.ID, *wr)
+		if err == nil || errors.Is(err, client.ErrRequestFailure) {
+			// we only retry if err is a request failure err.
+			return err
+		}
+
+		// otherwise retrying won't fix it, so we can terminate
+		return backoff.Permanent(err)
+	}, e.strategy)
 }
 
 // Deleted implements provision.Feedbacker
 func (e *Feedback) Deleted(nodeID, id string) error {
-	return e.client.Workloads.NodeWorkloadPutDeleted(nodeID, id)
+	return backoff.Retry(func() error {
+		err := e.client.Workloads.NodeWorkloadPutDeleted(nodeID, id)
+		if err == nil || errors.Is(err, client.ErrRequestFailure) {
+			// we only retry if err is a request failure err.
+			return err
+		}
+
+		// otherwise retrying won't fix it, so we can terminate
+		return backoff.Permanent(err)
+	}, e.strategy)
 }
 
 // UpdateStats implements provision.Feedbacker
 func (e *Feedback) UpdateStats(nodeID string, w directory.WorkloadAmount, u directory.ResourceAmount) error {
-	return e.client.Directory.NodeUpdateUsedResources(nodeID, u, w)
+	return backoff.Retry(func() error {
+		err := e.client.Directory.NodeUpdateUsedResources(nodeID, u, w)
+		if err == nil || errors.Is(err, client.ErrRequestFailure) {
+			// we only retry if err is a request failure err.
+			return err
+		}
+
+		// otherwise retrying won't fix it, so we can terminate
+		return backoff.Permanent(err)
+	}, e.strategy)
 }

--- a/pkg/stubs/provision_stub.go
+++ b/pkg/stubs/provision_stub.go
@@ -6,14 +6,14 @@ import (
 	pkg "github.com/threefoldtech/zos/pkg"
 )
 
-type ProvisionMonitorStub struct {
+type ProvisionStub struct {
 	client zbus.Client
 	module string
 	object zbus.ObjectID
 }
 
-func NewProvisionMonitorStub(client zbus.Client) *ProvisionMonitorStub {
-	return &ProvisionMonitorStub{
+func NewProvisionStub(client zbus.Client) *ProvisionStub {
+	return &ProvisionStub{
 		client: client,
 		module: "provision",
 		object: zbus.ObjectID{
@@ -23,7 +23,7 @@ func NewProvisionMonitorStub(client zbus.Client) *ProvisionMonitorStub {
 	}
 }
 
-func (s *ProvisionMonitorStub) Counters(ctx context.Context) (<-chan pkg.ProvisionCounters, error) {
+func (s *ProvisionStub) Counters(ctx context.Context) (<-chan pkg.ProvisionCounters, error) {
 	ch := make(chan pkg.ProvisionCounters)
 	recv, err := s.client.Stream(ctx, s.module, s.object, "Counters")
 	if err != nil {
@@ -40,4 +40,17 @@ func (s *ProvisionMonitorStub) Counters(ctx context.Context) (<-chan pkg.Provisi
 		}
 	}()
 	return ch, nil
+}
+
+func (s *ProvisionStub) DecommissionCached(arg0 string, arg1 string) (ret0 error) {
+	args := []interface{}{arg0, arg1}
+	result, err := s.client.Request(s.module, s.object, "DecommissionCached", args...)
+	if err != nil {
+		panic(err)
+	}
+	ret0 = new(zbus.RemoteError)
+	if err := result.Unmarshal(0, &ret0); err != nil {
+		panic(err)
+	}
+	return
 }

